File: examples/substrate/typescript/effect_boundary_server.ts
/**
 * HELM EffectBoundary Substrate Example — TypeScript
 *
 * Demonstrates implementing the EffectBoundary contract in TypeScript
 * using the OpenAPI-defined REST surface.
 */

import { createHash } from "crypto";
import { createServer, IncomingMessage, ServerResponse } from "http";

// --- Types ---

type Verdict = "ALLOW" | "DENY" | "ESCALATE";

interface EffectRequest {
  effect_type: string;
  principal_id: string;
  params?: Record<string, unknown>;
  context?: Record<string, unknown>;
  idempotency_key?: string;
}

interface Receipt {
  receipt_id: string;
  verdict: Verdict;
  reason_code: string;
  reason: string;
  timestamp: string;
  lamport: number;
  principal_id: string;
}

// --- EffectBoundary ---

class EffectBoundary {
  private lamport = 0;
  private receipts: Receipt[] = [];

  submit(req: EffectRequest): {
    verdict: Verdict;
    receipt: Receipt;
    intent: { effect_type: string; allowed: boolean };
  } {
    this.lamport++;
    const { verdict, reasonCode, reason } = this.evaluate(req);

    const receipt = this.createReceipt(
      verdict,
      reasonCode,
      reason,
      req.principal_id,
    );
    this.receipts.push(receipt);

    return {
      verdict,
      receipt,
      intent: {
        effect_type: req.effect_type,
        allowed: verdict === "ALLOW",
      },
    };
  }

  complete(receiptId: string, result: Record<string, unknown>): Receipt {
    this.lamport++;
    const receipt = this.createReceipt(
      "ALLOW",
      "EFFECT_COMPLETED",
      `Effect ${receiptId} completed`,
      "system",
    );
    this.receipts.push(receipt);
    return receipt;
  }

  private evaluate(req: EffectRequest): {
    verdict: Verdict;
    reasonCode: string;
    reason: string;
  } {
    if (req.effect_type === "data_export" && req.params?.data_class === "PII") {
      return {
        verdict: "DENY",
        reasonCode: "POLICY_VIOLATION",
        reason: "PII export denied by policy",
      };
    }

    if (req.effect_type === "financial_transfer") {
      const amount = (req.params?.amount_cents as number) ?? 0;
      if (amount > 1_000_000) {
        return {
          verdict: "ESCALATE",
          reasonCode: "TEMPORAL_INTERVENTION",
          reason: "High value transfer requires approval",
        };
      }
    }

    return {
      verdict: "ALLOW",
      reasonCode: "POLICY_SATISFIED",
      reason: "Effect allowed",
    };
  }

  private createReceipt(
    verdict: Verdict,
    reasonCode: string,
    reason: string,
    principalId: string,
  ): Receipt {
    const ts = new Date().toISOString();
    const content = `${verdict}:${reasonCode}:${ts}:${this.lamport}`;
    const hash = createHash("sha256")
      .update(content)
      .digest("hex")
      .slice(0, 16);

    return {
      receipt_id: `urn:helm:receipt:${hash}`,
      verdict,
      reason_code: reasonCode,
      reason,
      timestamp: ts,
      lamport: this.lamport,
      principal_id: principalId,
    };
  }
}

// --- HTTP Server ---

const boundary = new EffectBoundary();

function readBody(req: IncomingMessage): Promise<string> {
  return new Promise((resolve) => {
    let data = "";
    req.on("data", (chunk) => (data += chunk));
    req.on("end", () => resolve(data));
  });
}

const server = createServer(
  async (req: IncomingMessage, res: ServerResponse) => {
    if (req.method !== "POST") {
      res.writeHead(405);
      res.end(JSON.stringify({ error: "Method not allowed" }));
      return;
    }

    const body = JSON.parse(await readBody(req));

    if (req.url === "/v1/effects") {
      const result = boundary.submit(body as EffectRequest);
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify(result, null, 2));
    } else if (req.url?.match(/^\/v1\/effects\/[^/]+\/complete$/)) {
      const receiptId = req.url.split("/")[3];
      const result = boundary.complete(receiptId, body);
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify(result, null, 2));
    } else {
      res.writeHead(404);
      res.end(JSON.stringify({ error: "Not found" }));
    }
  },
);

console.log("HELM EffectBoundary substrate running on :4001");
server.listen(4001);

File: examples/substrate/python/effect_boundary_server.py
"""
HELM EffectBoundary Substrate Example — Python

Demonstrates implementing the EffectBoundary contract in Python
using the OpenAPI-defined REST surface.
"""

import json
import hashlib
import time
from dataclasses import dataclass, field, asdict
from enum import Enum
from typing import Optional
from http.server import HTTPServer, BaseHTTPRequestHandler


class Verdict(Enum):
    ALLOW = "ALLOW"
    DENY = "DENY"
    ESCALATE = "ESCALATE"


@dataclass
class EffectRequest:
    effect_type: str
    principal_id: str
    params: dict = field(default_factory=dict)
    context: dict = field(default_factory=dict)
    idempotency_key: Optional[str] = None


@dataclass
class Receipt:
    receipt_id: str
    verdict: str
    reason_code: str
    reason: str
    timestamp: str
    lamport: int
    principal_id: str

    @staticmethod
    def create(verdict: Verdict, reason_code: str, reason: str,
               principal_id: str, lamport: int) -> "Receipt":
        ts = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
        content = f"{verdict.value}:{reason_code}:{ts}:{lamport}"
        receipt_id = f"urn:helm:receipt:{hashlib.sha256(content.encode()).hexdigest()[:16]}"
        return Receipt(
            receipt_id=receipt_id,
            verdict=verdict.value,
            reason_code=reason_code,
            reason=reason,
            timestamp=ts,
            lamport=lamport,
            principal_id=principal_id,
        )


class EffectBoundary:
    """Minimal EffectBoundary implementation for substrate demonstration."""

    def __init__(self):
        self._lamport = 0
        self._receipts: list[Receipt] = []

    def submit(self, req: EffectRequest) -> dict:
        """Submit an effect for governance evaluation."""
        self._lamport += 1

        # --- PDP evaluation (simplified) ---
        verdict, reason_code, reason = self._evaluate(req)

        receipt = Receipt.create(
            verdict=verdict,
            reason_code=reason_code,
            reason=reason,
            principal_id=req.principal_id,
            lamport=self._lamport,
        )
        self._receipts.append(receipt)

        return {
            "verdict": verdict.value,
            "receipt": asdict(receipt),
            "intent": {
                "effect_type": req.effect_type,
                "allowed": verdict == Verdict.ALLOW,
            },
        }

    def complete(self, receipt_id: str, result: dict) -> dict:
        """Record completion of a previously allowed effect."""
        self._lamport += 1
        completion = Receipt.create(
            verdict=Verdict.ALLOW,
            reason_code="EFFECT_COMPLETED",
            reason=f"Effect {receipt_id} completed",
            principal_id="system",
            lamport=self._lamport,
        )
        self._receipts.append(completion)
        return asdict(completion)

    def _evaluate(self, req: EffectRequest) -> tuple[Verdict, str, str]:
        """Simplified PDP: deny risky effects, allow everything else."""
        if req.effect_type == "data_export" and req.params.get("data_class") == "PII":
            return Verdict.DENY, "POLICY_VIOLATION", "PII export denied by policy"
        if req.effect_type == "financial_transfer":
            amount = req.params.get("amount_cents", 0)
            if amount > 1_000_000:
                return Verdict.ESCALATE, "TEMPORAL_INTERVENTION", "High value transfer requires approval"
        return Verdict.ALLOW, "POLICY_SATISFIED", "Effect allowed"


class SubstrateHandler(BaseHTTPRequestHandler):
    """HTTP handler implementing the HELM EffectBoundary REST surface."""

    boundary = EffectBoundary()

    def do_POST(self):
        body = json.loads(self.rfile.read(int(self.headers["Content-Length"])))

        if self.path == "/v1/effects":
            req = EffectRequest(**body)
            result = self.boundary.submit(req)
            self._json_response(200, result)

        elif self.path.startswith("/v1/effects/") and self.path.endswith("/complete"):
            receipt_id = self.path.split("/")[3]
            result = self.boundary.complete(receipt_id, body)
            self._json_response(200, result)

        else:
            self._json_response(404, {"error": "Not found"})

    def _json_response(self, code: int, data: dict):
        self.send_response(code)
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(json.dumps(data, indent=2).encode())


if __name__ == "__main__":
    print("HELM EffectBoundary substrate running on :4001")
    HTTPServer(("", 4001), SubstrateHandler).serve_forever()
